plugin/pulpcore/plugin/stages/content_unit_stages.py

```diff
@@ -0,0 +1,176 @@
+    inspects their associated :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object stores one
+    :class:`~pulpcore.plugin.models.Artifact`.
+
+    This stage inspects any "unsaved" Content unit objects and searches for existing saved Content
+    units inside Pulp with the same unit key. Any existing Content objects found, replace their
+    "unsaved" counterpart in the :class:`~pulpcore.plugin.stages.DeclarativeContent` object.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after it has been
+    handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q (:class:`asyncio.Queue`): Each item is a
+            :class:`~pulpcore.plugin.stages.DeclarativeContent`
+        out_q (:class:`asyncio.Queue`): Each item is a
+            :class:`~pulpcore.plugin.stages.DeclarativeContent`
+
+    Returns:
+        The query_existing_content_units stage as a coroutine to be included in a pipeline.
+    """
+    declarative_content_list = []
+    shutdown = False
+    while True:
+        try:
+            content = in_q.get_nowait()
+        except asyncio.QueueEmpty:
+            if not declarative_content_list:
+                content = await in_q.get()
+                declarative_content_list.append(content)
+                continue
+        else:
+            declarative_content_list.append(content)
+            continue
+
+        content_q_by_type = defaultdict(lambda: Q(pk=None))
+        for declarative_content in declarative_content_list:
+            if declarative_content is None:
+                shutdown = True
+                continue
+
+            model_type = type(declarative_content.content)
+            unit_key = declarative_content.content.natural_key_dict()
+            content_q_by_type[model_type] = content_q_by_type[model_type] | Q(**unit_key)
+
+        for model_type in content_q_by_type.keys():
+            for result in model_type.objects.filter(content_q_by_type[model_type]):
+                for declarative_content in declarative_content_list:
+                    if declarative_content is None:
+                        continue
+                    not_same_unit = False
+                    for field in result.natural_key_fields():
+                        if getattr(declarative_content.content, field) != getattr(result, field):
+                            not_same_unit = True
+                            break
+                    if not_same_unit:
+                        continue
+                    declarative_content.content = result
+        for declarative_content in declarative_content_list:
+            await out_q.put(declarative_content)
+        declarative_content_list = []
+        if shutdown:
+            break
+    await out_q.put(None)
+
+
+async def content_unit_saver(in_q, out_q):
```

Comment:
This function contains multiple code blocks each doing something complex.  Recommend decomposing into smaller functions each doing *one* thing.  Better functional decomposition will be more readable, testable and maintainable.

Reply:
I agree, but I hope we can do this later. See similar answer [here](https://github.com/pulp/pulp/pull/3559#discussion_r206947935).